### ISSUE TITLE
Report generation OOME

### ISSUE THREAD

**Issue by ghost** [testng-team/testng#533]
Using version 6.8.8:

```
java.lang.OutOfMemoryError: Java heap space
        at java.util.Arrays.copyOf(Arrays.java:2367)
        at java.lang.AbstractStringBuilder.expandCapacity(AbstractStringBuilder.java:130)
        at java.lang.AbstractStringBuilder.ensureCapacityInternal(AbstractStringBuilder.java:114)
        at java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:415)
        at java.lang.StringBuilder.append(StringBuilder.java:132)
        at org.testng.reporters.Files.readFile(Files.java:28)
        at org.testng.reporters.Files.readFile(Files.java:20)
        at org.testng.reporters.FileStringBuffer.toString(FileStringBuffer.java:137)
        at org.testng.reporters.XMLStringBuffer.toXML(XMLStringBuffer.java:331)
        at org.testng.reporters.jq.ChronologicalPanel.getContent(ChronologicalPanel.java:93)
        at org.testng.reporters.jq.BaseMultiSuitePanel.generate(BaseMultiSuitePanel.java:24)
        at org.testng.reporters.jq.Main.generateReport(Main.java:63)
        at org.testng.TestNG.generateReports(TestNG.java:1115)
        at org.testng.TestNG.run(TestNG.java:1074)
        at org.apache.maven.surefire.testng.TestNGExecutor.run(TestNGExecutor.java:115)
        at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.executeMulti(TestNGDirectoryTestSuite.java:204)
        at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.execute(TestNGDirectoryTestSuite.java:107)
        at org.apache.maven.surefire.testng.TestNGProvider.invoke(TestNGProvider.java:112)
        at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:200)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:153)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:103)
```

The exception appears when there are a more than a few test failures.

**Comment by SlavikCA** [comment]
Looks similar to this:
https://github.com/cbeust/testng/issues/291

It says, that it fixed in latest version, but I still has similar issue with 6.8.13

The only workaround I found, is to clear Report regularly:
```java
try {
    assert compareScreenShots(frame, actual, "build/screenshots/proxy"), 
      "Too many differences between the two panes, this URL fails: ${url}"
}
finally {
    Reporter.clear()
}
```

Somehow, at the end report generated anyway

**Comment by SlavikCA** [comment]
I found, that my advice above only partially resolve issue - memory still grows, just slower... At the end, I still getting OutOfMemory

**Comment by cbeust** [comment]
It's odd that `Reporter.clear()` would help: are you logging a lot of things in the Reporter? Do you have a lot of tests? Can you break point into `Reporter.clear()` and see how big the two fields that get cleared are?

**Comment by SlavikCA** [comment]
I do not log anything explicitly, But I found, that testNG default Reporter logs all test case properties, as I can see in Heap Walker.
So, content of screenshots are written in the Reporter - that's a lot of data. And I do not know, how I can make it NOT write this data in the Reporter. I tried to set them to null, but it didn't work.

Anyway, I tried to debug, as you suggested, and found one strange thing, which would explain, why even with Reporter.clear() memory still grows:
The code in my "finally" block gets executed ONLY if assert PASS. If it FAIL - than code in "finally" block never gets to execute. Is it intended? It certainly not intuitive.

Here is Heap Walker report:
http://slavikf.com/files/testNG1/

Note, that  java.util.HashMap$Node element created for each test run (I'm using data provider with array of URLs) and each of that element contain 2 screenshots (frameImg and actualImg). So, it grows out-of-control.
And it seems, that Reporter.clear() takes care of it.
It just, that now I realize, that I need to put it after Assert, too.

**Comment by cbeust** [comment]
Indeed, thanks for the heap dump.

I'm guessing you are using Selenium and adding screen shots in the test result, since TestNG itself doesn't use the Reporter at all.

I think the main problem here is that `ITestResult` is used as a key in the `m_methodOutputMap`, and this is really not necessary at all: all I need is a unique identifier from a test result in order to get the logged data. This would remove all the `ITestResult` from the maps and replace them with Strings (most likely) as keys.

Actually, even using the `ITestResult`'s hashCode() should be good enough, since it seems it's already doing that (and using the default one, so identityHashCode really).

I can create a patched testng.jar if you're interested, you can test it and let me know if it helps.

Please email me directly and I'll send it to you.

Thanks!

**Comment by krmahadevan** [comment]
Reporter class now uses only the method id as the key for `m_methodOutputMap`. Please comment/add a sample that can be used to reproduce the problem if issue still exists in latest released version of TestNG as of today `7.8.0`
